// cashc/dist/grammar/CashScriptParser.js
// Generated from src/grammar/CashScript.g4 by ANTLR 4.13.1
// noinspection ES6UnusedImports,JSUnusedGlobalSymbols,JSUnusedLocalSymbols
import { ATN, ATNDeserializer, DFA, FailedPredicateException, RecognitionException, NoViableAltException, Parser, ParserATNSimulator, ParserRuleContext, PredictionContextCache, Token } from 'antlr4';
class CashScriptParser extends Parser {
    get grammarFileName() { return "CashScript.g4"; }
    get literalNames() { return CashScriptParser.literalNames; }
    get symbolicNames() { return CashScriptParser.symbolicNames; }
    get ruleNames() { return CashScriptParser.ruleNames; }
    get serializedATN() { return CashScriptParser._serializedATN; }
    createFailedPredicateException(predicate, message) {
        return new FailedPredicateException(this, predicate, message);
    }
    constructor(input) {
        super(input);
        this._interp = new ParserATNSimulator(this, CashScriptParser._ATN, CashScriptParser.DecisionsToDFA, new PredictionContextCache());
    }
    // @RuleVersion(0)
    sourceFile() {
        let localctx = new SourceFileContext(this, this._ctx, this.state);
        this.enterRule(localctx, 0, CashScriptParser.RULE_sourceFile);
        let _la;
        try {
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 61;
                this._errHandler.sync(this);
                _la = this._input.LA(1);
                while (_la === 1) {
                    {
                        {
                            this.state = 58;
                            this.pragmaDirective();
                        }
                    }
                    this.state = 63;
                    this._errHandler.sync(this);
                    _la = this._input.LA(1);
                }
                this.state = 64;
                this.contractDefinition();
                this.state = 65;
                this.match(CashScriptParser.EOF);
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    pragmaDirective() {
        let localctx = new PragmaDirectiveContext(this, this._ctx, this.state);
        this.enterRule(localctx, 2, CashScriptParser.RULE_pragmaDirective);
        try {
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 67;
                this.match(CashScriptParser.T__0);
                this.state = 68;
                this.pragmaName();
                this.state = 69;
                this.pragmaValue();
                this.state = 70;
                this.match(CashScriptParser.T__1);
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    pragmaName() {
        let localctx = new PragmaNameContext(this, this._ctx, this.state);
        this.enterRule(localctx, 4, CashScriptParser.RULE_pragmaName);
        try {
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 72;
                this.match(CashScriptParser.T__2);
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    pragmaValue() {
        let localctx = new PragmaValueContext(this, this._ctx, this.state);
        this.enterRule(localctx, 6, CashScriptParser.RULE_pragmaValue);
        let _la;
        try {
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 74;
                this.versionConstraint();
                this.state = 76;
                this._errHandler.sync(this);
                _la = this._input.LA(1);
                if ((((_la) & ~0x1F) === 0 && ((1 << _la) & 2032) !== 0) || _la === 59) {
                    {
                        this.state = 75;
                        this.versionConstraint();
                    }
                }
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    versionConstraint() {
        let localctx = new VersionConstraintContext(this, this._ctx, this.state);
        this.enterRule(localctx, 8, CashScriptParser.RULE_versionConstraint);
        let _la;
        try {
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 79;
                this._errHandler.sync(this);
                _la = this._input.LA(1);
                if ((((_la) & ~0x1F) === 0 && ((1 << _la) & 2032) !== 0)) {
                    {
                        this.state = 78;
                        this.versionOperator();
                    }
                }
                this.state = 81;
                this.match(CashScriptParser.VersionLiteral);
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    versionOperator() {
        let localctx = new VersionOperatorContext(this, this._ctx, this.state);
        this.enterRule(localctx, 10, CashScriptParser.RULE_versionOperator);
        let _la;
        try {
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 83;
                _la = this._input.LA(1);
                if (!((((_la) & ~0x1F) === 0 && ((1 << _la) & 2032) !== 0))) {
                    this._errHandler.recoverInline(this);
                }
                else {
                    this._errHandler.reportMatch(this);
                    this.consume();
                }
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    contractDefinition() {
        let localctx = new ContractDefinitionContext(this, this._ctx, this.state);
        this.enterRule(localctx, 12, CashScriptParser.RULE_contractDefinition);
        let _la;
        try {
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 85;
                this.match(CashScriptParser.T__10);
                this.state = 86;
                this.match(CashScriptParser.Identifier);
                this.state = 87;
                this.parameterList();
                this.state = 88;
                this.match(CashScriptParser.T__11);
                this.state = 92;
                this._errHandler.sync(this);
                _la = this._input.LA(1);
                while (_la === 14) {
                    {
                        {
                            this.state = 89;
                            this.functionDefinition();
                        }
                    }
                    this.state = 94;
                    this._errHandler.sync(this);
                    _la = this._input.LA(1);
                }
                this.state = 95;
                this.match(CashScriptParser.T__12);
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    functionDefinition() {
        let localctx = new FunctionDefinitionContext(this, this._ctx, this.state);
        this.enterRule(localctx, 14, CashScriptParser.RULE_functionDefinition);
        let _la;
        try {
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 97;
                this.match(CashScriptParser.T__13);
                this.state = 98;
                this.match(CashScriptParser.Identifier);
                this.state = 99;
                this.parameterList();
                this.state = 100;
                this.match(CashScriptParser.T__11);
                this.state = 104;
                this._errHandler.sync(this);
                _la = this._input.LA(1);
                while ((((_la) & ~0x1F) === 0 && ((1 << _la) & 2883584) !== 0) || ((((_la - 53)) & ~0x1F) === 0 && ((1 << (_la - 53)) & 528447) !== 0)) {
                    {
                        {
                            this.state = 101;
                            this.statement();
                        }
                    }
                    this.state = 106;
                    this._errHandler.sync(this);
                    _la = this._input.LA(1);
                }
                this.state = 107;
                this.match(CashScriptParser.T__12);
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    parameterList() {
        let localctx = new ParameterListContext(this, this._ctx, this.state);
        this.enterRule(localctx, 16, CashScriptParser.RULE_parameterList);
        let _la;
        try {
            let _alt;
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 109;
                this.match(CashScriptParser.T__14);
                this.state = 121;
                this._errHandler.sync(this);
                _la = this._input.LA(1);
                if (((((_la - 53)) & ~0x1F) === 0 && ((1 << (_la - 53)) & 4159) !== 0)) {
                    {
                        this.state = 110;
                        this.parameter();
                        this.state = 115;
                        this._errHandler.sync(this);
                        _alt = this._interp.adaptivePredict(this._input, 5, this._ctx);
                        while (_alt !== 2 && _alt !== ATN.INVALID_ALT_NUMBER) {
                            if (_alt === 1) {
                                {
                                    {
                                        this.state = 111;
                                        this.match(CashScriptParser.T__15);
                                        this.state = 112;
                                        this.parameter();
                                    }
                                }
                            }
                            this.state = 117;
                            this._errHandler.sync(this);
                            _alt = this._interp.adaptivePredict(this._input, 5, this._ctx);
                        }
                        this.state = 119;
                        this._errHandler.sync(this);
                        _la = this._input.LA(1);
                        if (_la === 16) {
                            {
                                this.state = 118;
                                this.match(CashScriptParser.T__15);
                            }
                        }
                    }
                }
                this.state = 123;
                this.match(CashScriptParser.T__16);
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    parameter() {
        let localctx = new ParameterContext(this, this._ctx, this.state);
        this.enterRule(localctx, 18, CashScriptParser.RULE_parameter);
        try {
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 125;
                this.typeName();
                this.state = 126;
                this.match(CashScriptParser.Identifier);
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    block() {
        let localctx = new BlockContext(this, this._ctx, this.state);
        this.enterRule(localctx, 20, CashScriptParser.RULE_block);
        let _la;
        try {
            this.state = 137;
            this._errHandler.sync(this);
            switch (this._input.LA(1)) {
                case 12:
                    this.enterOuterAlt(localctx, 1);
                    {
                        this.state = 128;
                        this.match(CashScriptParser.T__11);
                        this.state = 132;
                        this._errHandler.sync(this);
                        _la = this._input.LA(1);
                        while ((((_la) & ~0x1F) === 0 && ((1 << _la) & 2883584) !== 0) || ((((_la - 53)) & ~0x1F) === 0 && ((1 << (_la - 53)) & 528447) !== 0)) {
                            {
                                {
                                    this.state = 129;
                                    this.statement();
                                }
                            }
                            this.state = 134;
                            this._errHandler.sync(this);
                            _la = this._input.LA(1);
                        }
                        this.state = 135;
                        this.match(CashScriptParser.T__12);
                    }
                    break;
                case 18:
                case 19:
                case 21:
                case 53:
                case 54:
                case 55:
                case 56:
                case 57:
                case 58:
                case 65:
                case 72:
                    this.enterOuterAlt(localctx, 2);
                    {
                        this.state = 136;
                        this.statement();
                    }
                    break;
                default:
                    throw new NoViableAltException(this);
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    statement() {
        let localctx = new StatementContext(this, this._ctx, this.state);
        this.enterRule(localctx, 22, CashScriptParser.RULE_statement);
        try {
            this.state = 146;
            this._errHandler.sync(this);
            switch (this._interp.adaptivePredict(this._input, 10, this._ctx)) {
                case 1:
                    this.enterOuterAlt(localctx, 1);
                    {
                        this.state = 139;
                        this.variableDefinition();
                    }
                    break;
                case 2:
                    this.enterOuterAlt(localctx, 2);
                    {
                        this.state = 140;
                        this.tupleAssignment();
                    }
                    break;
                case 3:
                    this.enterOuterAlt(localctx, 3);
                    {
                        this.state = 141;
                        this.assignStatement();
                    }
                    break;
                case 4:
                    this.enterOuterAlt(localctx, 4);
                    {
                        this.state = 142;
                        this.timeOpStatement();
                    }
                    break;
                case 5:
                    this.enterOuterAlt(localctx, 5);
                    {
                        this.state = 143;
                        this.requireStatement();
                    }
                    break;
                case 6:
                    this.enterOuterAlt(localctx, 6);
                    {
                        this.state = 144;
                        this.ifStatement();
                    }
                    break;
                case 7:
                    this.enterOuterAlt(localctx, 7);
                    {
                        this.state = 145;
                        this.consoleStatement();
                    }
                    break;
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    variableDefinition() {
        let localctx = new VariableDefinitionContext(this, this._ctx, this.state);
        this.enterRule(localctx, 24, CashScriptParser.RULE_variableDefinition);
        let _la;
        try {
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 148;
                this.typeName();
                this.state = 152;
                this._errHandler.sync(this);
                _la = this._input.LA(1);
                while (_la === 52) {
                    {
                        {
                            this.state = 149;
                            this.modifier();
                        }
                    }
                    this.state = 154;
                    this._errHandler.sync(this);
                    _la = this._input.LA(1);
                }
                this.state = 155;
                this.match(CashScriptParser.Identifier);
                this.state = 156;
                this.match(CashScriptParser.T__9);
                this.state = 157;
                this.expression(0);
                this.state = 158;
                this.match(CashScriptParser.T__1);
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    tupleAssignment() {
        let localctx = new TupleAssignmentContext(this, this._ctx, this.state);
        this.enterRule(localctx, 26, CashScriptParser.RULE_tupleAssignment);
        try {
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 160;
                this.typeName();
                this.state = 161;
                this.match(CashScriptParser.Identifier);
                this.state = 162;
                this.match(CashScriptParser.T__15);
                this.state = 163;
                this.typeName();
                this.state = 164;
                this.match(CashScriptParser.Identifier);
                this.state = 165;
                this.match(CashScriptParser.T__9);
                this.state = 166;
                this.expression(0);
                this.state = 167;
                this.match(CashScriptParser.T__1);
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    assignStatement() {
        let localctx = new AssignStatementContext(this, this._ctx, this.state);
        this.enterRule(localctx, 28, CashScriptParser.RULE_assignStatement);
        try {
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 169;
                this.match(CashScriptParser.Identifier);
                this.state = 170;
                this.match(CashScriptParser.T__9);
                this.state = 171;
                this.expression(0);
                this.state = 172;
                this.match(CashScriptParser.T__1);
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    timeOpStatement() {
        let localctx = new TimeOpStatementContext(this, this._ctx, this.state);
        this.enterRule(localctx, 30, CashScriptParser.RULE_timeOpStatement);
        let _la;
        try {
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 174;
                this.match(CashScriptParser.T__17);
                this.state = 175;
                this.match(CashScriptParser.T__14);
                this.state = 176;
                this.match(CashScriptParser.TxVar);
                this.state = 177;
                this.match(CashScriptParser.T__5);
                this.state = 178;
                this.expression(0);
                this.state = 181;
                this._errHandler.sync(this);
                _la = this._input.LA(1);
                if (_la === 16) {
                    {
                        this.state = 179;
                        this.match(CashScriptParser.T__15);
                        this.state = 180;
                        this.requireMessage();
                    }
                }
                this.state = 183;
                this.match(CashScriptParser.T__16);
                this.state = 184;
                this.match(CashScriptParser.T__1);
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    requireStatement() {
        let localctx = new RequireStatementContext(this, this._ctx, this.state);
        this.enterRule(localctx, 32, CashScriptParser.RULE_requireStatement);
        let _la;
        try {
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 186;
                this.match(CashScriptParser.T__17);
                this.state = 187;
                this.match(CashScriptParser.T__14);
                this.state = 188;
                this.expression(0);
                this.state = 191;
                this._errHandler.sync(this);
                _la = this._input.LA(1);
                if (_la === 16) {
                    {
                        this.state = 189;
                        this.match(CashScriptParser.T__15);
                        this.state = 190;
                        this.requireMessage();
                    }
                }
                this.state = 193;
                this.match(CashScriptParser.T__16);
                this.state = 194;
                this.match(CashScriptParser.T__1);
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    ifStatement() {
        let localctx = new IfStatementContext(this, this._ctx, this.state);
        this.enterRule(localctx, 34, CashScriptParser.RULE_ifStatement);
        try {
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 196;
                this.match(CashScriptParser.T__18);
                this.state = 197;
                this.match(CashScriptParser.T__14);
                this.state = 198;
                this.expression(0);
                this.state = 199;
                this.match(CashScriptParser.T__16);
                this.state = 200;
                localctx._ifBlock = this.block();
                this.state = 203;
                this._errHandler.sync(this);
                switch (this._interp.adaptivePredict(this._input, 14, this._ctx)) {
                    case 1:
                        {
                            this.state = 201;
                            this.match(CashScriptParser.T__19);
                            this.state = 202;
                            localctx._elseBlock = this.block();
                        }
                        break;
                }
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    consoleStatement() {
        let localctx = new ConsoleStatementContext(this, this._ctx, this.state);
        this.enterRule(localctx, 36, CashScriptParser.RULE_consoleStatement);
        try {
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 205;
                this.match(CashScriptParser.T__20);
                this.state = 206;
                this.consoleParameterList();
                this.state = 207;
                this.match(CashScriptParser.T__1);
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    requireMessage() {
        let localctx = new RequireMessageContext(this, this._ctx, this.state);
        this.enterRule(localctx, 38, CashScriptParser.RULE_requireMessage);
        try {
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 209;
                this.match(CashScriptParser.StringLiteral);
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    consoleParameter() {
        let localctx = new ConsoleParameterContext(this, this._ctx, this.state);
        this.enterRule(localctx, 40, CashScriptParser.RULE_consoleParameter);
        try {
            this.state = 213;
            this._errHandler.sync(this);
            switch (this._input.LA(1)) {
                case 72:
                    this.enterOuterAlt(localctx, 1);
                    {
                        this.state = 211;
                        this.match(CashScriptParser.Identifier);
                    }
                    break;
                case 60:
                case 62:
                case 67:
                case 68:
                case 69:
                    this.enterOuterAlt(localctx, 2);
                    {
                        this.state = 212;
                        this.literal();
                    }
                    break;
                default:
                    throw new NoViableAltException(this);
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    consoleParameterList() {
        let localctx = new ConsoleParameterListContext(this, this._ctx, this.state);
        this.enterRule(localctx, 42, CashScriptParser.RULE_consoleParameterList);
        let _la;
        try {
            let _alt;
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 215;
                this.match(CashScriptParser.T__14);
                this.state = 227;
                this._errHandler.sync(this);
                _la = this._input.LA(1);
                if (((((_la - 60)) & ~0x1F) === 0 && ((1 << (_la - 60)) & 4997) !== 0)) {
                    {
                        this.state = 216;
                        this.consoleParameter();
                        this.state = 221;
                        this._errHandler.sync(this);
                        _alt = this._interp.adaptivePredict(this._input, 16, this._ctx);
                        while (_alt !== 2 && _alt !== ATN.INVALID_ALT_NUMBER) {
                            if (_alt === 1) {
                                {
                                    {
                                        this.state = 217;
                                        this.match(CashScriptParser.T__15);
                                        this.state = 218;
                                        this.consoleParameter();
                                    }
                                }
                            }
                            this.state = 223;
                            this._errHandler.sync(this);
                            _alt = this._interp.adaptivePredict(this._input, 16, this._ctx);
                        }
                        this.state = 225;
                        this._errHandler.sync(this);
                        _la = this._input.LA(1);
                        if (_la === 16) {
                            {
                                this.state = 224;
                                this.match(CashScriptParser.T__15);
                            }
                        }
                    }
                }
                this.state = 229;
                this.match(CashScriptParser.T__16);
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    functionCall() {
        let localctx = new FunctionCallContext(this, this._ctx, this.state);
        this.enterRule(localctx, 44, CashScriptParser.RULE_functionCall);
        try {
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 231;
                this.match(CashScriptParser.Identifier);
                this.state = 232;
                this.expressionList();
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    expressionList() {
        let localctx = new ExpressionListContext(this, this._ctx, this.state);
        this.enterRule(localctx, 46, CashScriptParser.RULE_expressionList);
        let _la;
        try {
            let _alt;
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 234;
                this.match(CashScriptParser.T__14);
                this.state = 246;
                this._errHandler.sync(this);
                _la = this._input.LA(1);
                if (((((_la - 15)) & ~0x1F) === 0 && ((1 << (_la - 15)) & 100730241) !== 0) || ((((_la - 53)) & ~0x1F) === 0 && ((1 << (_la - 53)) & 905919) !== 0)) {
                    {
                        this.state = 235;
                        this.expression(0);
                        this.state = 240;
                        this._errHandler.sync(this);
                        _alt = this._interp.adaptivePredict(this._input, 19, this._ctx);
                        while (_alt !== 2 && _alt !== ATN.INVALID_ALT_NUMBER) {
                            if (_alt === 1) {
                                {
                                    {
                                        this.state = 236;
                                        this.match(CashScriptParser.T__15);
                                        this.state = 237;
                                        this.expression(0);
                                    }
                                }
                            }
                            this.state = 242;
                            this._errHandler.sync(this);
                            _alt = this._interp.adaptivePredict(this._input, 19, this._ctx);
                        }
                        this.state = 244;
                        this._errHandler.sync(this);
                        _la = this._input.LA(1);
                        if (_la === 16) {
                            {
                                this.state = 243;
                                this.match(CashScriptParser.T__15);
                            }
                        }
                    }
                }
                this.state = 248;
                this.match(CashScriptParser.T__16);
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    expression(_p) {
        if (_p === undefined) {
            _p = 0;
        }
        let _parentctx = this._ctx;
        let _parentState = this.state;
        let localctx = new ExpressionContext(this, this._ctx, _parentState);
        let _prevctx = localctx;
        let _startState = 48;
        this.enterRecursionRule(localctx, 48, CashScriptParser.RULE_expression, _p);
        let _la;
        try {
            let _alt;
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 303;
                this._errHandler.sync(this);
                switch (this._interp.adaptivePredict(this._input, 27, this._ctx)) {
                    case 1:
                        {
                            localctx = new ParenthesisedContext(this, localctx);
                            this._ctx = localctx;
                            _prevctx = localctx;
                            this.state = 251;
                            this.match(CashScriptParser.T__14);
                            this.state = 252;
                            this.expression(0);
                            this.state = 253;
                            this.match(CashScriptParser.T__16);
                        }
                        break;
                    case 2:
                        {
                            localctx = new CastContext(this, localctx);
                            this._ctx = localctx;
                            _prevctx = localctx;
                            this.state = 255;
                            this.typeName();
                            this.state = 256;
                            this.match(CashScriptParser.T__14);
                            this.state = 257;
                            localctx._castable = this.expression(0);
                            this.state = 260;
                            this._errHandler.sync(this);
                            switch (this._interp.adaptivePredict(this._input, 22, this._ctx)) {
                                case 1:
                                    {
                                        this.state = 258;
                                        this.match(CashScriptParser.T__15);
                                        this.state = 259;
                                        localctx._size = this.expression(0);
                                    }
                                    break;
                            }
                            this.state = 263;
                            this._errHandler.sync(this);
                            _la = this._input.LA(1);
                            if (_la === 16) {
                                {
                                    this.state = 262;
                                    this.match(CashScriptParser.T__15);
                                }
                            }
                            this.state = 265;
                            this.match(CashScriptParser.T__16);
                        }
                        break;
                    case 3:
                        {
                            localctx = new FunctionCallExpressionContext(this, localctx);
                            this._ctx = localctx;
                            _prevctx = localctx;
                            this.state = 267;
                            this.functionCall();
                        }
                        break;
                    case 4:
                        {
                            localctx = new InstantiationContext(this, localctx);
                            this._ctx = localctx;
                            _prevctx = localctx;
                            this.state = 268;
                            this.match(CashScriptParser.T__21);
                            this.state = 269;
                            this.match(CashScriptParser.Identifier);
                            this.state = 270;
                            this.expressionList();
                        }
                        break;
                    case 5:
                        {
                            localctx = new UnaryIntrospectionOpContext(this, localctx);
                            this._ctx = localctx;
                            _prevctx = localctx;
                            this.state = 271;
                            localctx._scope = this.match(CashScriptParser.T__24);
                            this.state = 272;
                            this.match(CashScriptParser.T__22);
                            this.state = 273;
                            this.expression(0);
                            this.state = 274;
                            this.match(CashScriptParser.T__23);
                            this.state = 275;
                            localctx._op = this._input.LT(1);
                            _la = this._input.LA(1);
                            if (!((((_la) & ~0x1F) === 0 && ((1 << _la) & 2080374784) !== 0))) {
                                localctx._op = this._errHandler.recoverInline(this);
                            }
                            else {
                                this._errHandler.reportMatch(this);
                                this.consume();
                            }
                        }
                        break;
                    case 6:
                        {
                            localctx = new UnaryIntrospectionOpContext(this, localctx);
                            this._ctx = localctx;
                            _prevctx = localctx;
                            this.state = 277;
                            localctx._scope = this.match(CashScriptParser.T__30);
                            this.state = 278;
                            this.match(CashScriptParser.T__22);
                            this.state = 279;
                            this.expression(0);
                            this.state = 280;
                            this.match(CashScriptParser.T__23);
                            this.state = 281;
                            localctx._op = this._input.LT(1);
                            _la = this._input.LA(1);
                            if (!(((((_la - 26)) & ~0x1F) === 0 && ((1 << (_la - 26)) & 991) !== 0))) {
                                localctx._op = this._errHandler.recoverInline(this);
                            }
                            else {
                                this._errHandler.reportMatch(this);
                                this.consume();
                            }
                        }
                        break;
                    case 7:
                        {
                            localctx = new UnaryOpContext(this, localctx);
                            this._ctx = localctx;
                            _prevctx = localctx;
                            this.state = 283;
                            localctx._op = this._input.LT(1);
                            _la = this._input.LA(1);
                            if (!(_la === 40 || _la === 41)) {
                                localctx._op = this._errHandler.recoverInline(this);
                            }
                            else {
                                this._errHandler.reportMatch(this);
                                this.consume();
                            }
                            this.state = 284;
                            this.expression(14);
                        }
                        break;
                    case 8:
                        {
                            localctx = new ArrayContext(this, localctx);
                            this._ctx = localctx;
                            _prevctx = localctx;
                            this.state = 285;
                            this.match(CashScriptParser.T__22);
                            this.state = 297;
                            this._errHandler.sync(this);
                            _la = this._input.LA(1);
                            if (((((_la - 15)) & ~0x1F) === 0 && ((1 << (_la - 15)) & 100730241) !== 0) || ((((_la - 53)) & ~0x1F) === 0 && ((1 << (_la - 53)) & 905919) !== 0)) {
                                {
                                    this.state = 286;
                                    this.expression(0);
                                    this.state = 291;
                                    this._errHandler.sync(this);
                                    _alt = this._interp.adaptivePredict(this._input, 24, this._ctx);
                                    while (_alt !== 2 && _alt !== ATN.INVALID_ALT_NUMBER) {
                                        if (_alt === 1) {
                                            {
                                                {
                                                    this.state = 287;
                                                    this.match(CashScriptParser.T__15);
                                                    this.state = 288;
                                                    this.expression(0);
                                                }
                                            }
                                        }
                                        this.state = 293;
                                        this._errHandler.sync(this);
                                        _alt = this._interp.adaptivePredict(this._input, 24, this._ctx);
                                    }
                                    this.state = 295;
                                    this._errHandler.sync(this);
                                    _la = this._input.LA(1);
                                    if (_la === 16) {
                                        {
                                            this.state = 294;
                                            this.match(CashScriptParser.T__15);
                                        }
                                    }
                                }
                            }
                            this.state = 299;
                            this.match(CashScriptParser.T__23);
                        }
                        break;
                    case 9:
                        {
                            localctx = new NullaryOpContext(this, localctx);
                            this._ctx = localctx;
                            _prevctx = localctx;
                            this.state = 300;
                            this.match(CashScriptParser.NullaryOp);
                        }
                        break;
                    case 10:
                        {
                            localctx = new IdentifierContext(this, localctx);
                            this._ctx = localctx;
                            _prevctx = localctx;
                            this.state = 301;
                            this.match(CashScriptParser.Identifier);
                        }
                        break;
                    case 11:
                        {
                            localctx = new LiteralExpressionContext(this, localctx);
                            this._ctx = localctx;
                            _prevctx = localctx;
                            this.state = 302;
                            this.literal();
                        }
                        break;
                }
                this._ctx.stop = this._input.LT(-1);
                this.state = 354;
                this._errHandler.sync(this);
                _alt = this._interp.adaptivePredict(this._input, 29, this._ctx);
                while (_alt !== 2 && _alt !== ATN.INVALID_ALT_NUMBER) {
                    if (_alt === 1) {
                        if (this._parseListeners != null) {
                            this.triggerExitRuleEvent();
                        }
                        _prevctx = localctx;
                        {
                            this.state = 352;
                            this._errHandler.sync(this);
                            switch (this._interp.adaptivePredict(this._input, 28, this._ctx)) {
                                case 1:
                                    {
                                        localctx = new BinaryOpContext(this, new ExpressionContext(this, _parentctx, _parentState));
                                        localctx._left = _prevctx;
                                        this.pushNewRecursionContext(localctx, _startState, CashScriptParser.RULE_expression);
                                        this.state = 305;
                                        if (!(this.precpred(this._ctx, 13))) {
                                            throw this.createFailedPredicateException("this.precpred(this._ctx, 13)");
                                        }
                                        this.state = 306;
                                        localctx._op = this._input.LT(1);
                                        _la = this._input.LA(1);
                                        if (!(((((_la - 42)) & ~0x1F) === 0 && ((1 << (_la - 42)) & 7) !== 0))) {
                                            localctx._op = this._errHandler.recoverInline(this);
                                        }
                                        else {
                                            this._errHandler.reportMatch(this);
                                            this.consume();
                                        }
                                        this.state = 307;
                                        localctx._right = this.expression(14);
                                    }
                                    break;
                                case 2:
                                    {
                                        localctx = new BinaryOpContext(this, new ExpressionContext(this, _parentctx, _parentState));
                                        localctx._left = _prevctx;
                                        this.pushNewRecursionContext(localctx, _startState, CashScriptParser.RULE_expression);
                                        this.state = 308;
                                        if (!(this.precpred(this._ctx, 12))) {
                                            throw this.createFailedPredicateException("this.precpred(this._ctx, 12)");
                                        }
                                        this.state = 309;
                                        localctx._op = this._input.LT(1);
                                        _la = this._input.LA(1);
                                        if (!(_la === 41 || _la === 45)) {
                                            localctx._op = this._errHandler.recoverInline(this);
                                        }
                                        else {
                                            this._errHandler.reportMatch(this);
                                            this.consume();
                                        }
                                        this.state = 310;
                                        localctx._right = this.expression(13);
                                    }
                                    break;
                                case 3:
                                    {
                                        localctx = new BinaryOpContext(this, new ExpressionContext(this, _parentctx, _parentState));
                                        localctx._left = _prevctx;
                                        this.pushNewRecursionContext(localctx, _startState, CashScriptParser.RULE_expression);
                                        this.state = 311;
                                        if (!(this.precpred(this._ctx, 11))) {
                                            throw this.createFailedPredicateException("this.precpred(this._ctx, 11)");
                                        }
                                        this.state = 312;
                                        localctx._op = this._input.LT(1);
                                        _la = this._input.LA(1);
                                        if (!((((_la) & ~0x1F) === 0 && ((1 << _la) & 960) !== 0))) {
                                            localctx._op = this._errHandler.recoverInline(this);
                                        }
                                        else {
                                            this._errHandler.reportMatch(this);
                                            this.consume();
                                        }
                                        this.state = 313;
                                        localctx._right = this.expression(12);
                                    }
                                    break;
                                case 4:
                                    {
                                        localctx = new BinaryOpContext(this, new ExpressionContext(this, _parentctx, _parentState));
                                        localctx._left = _prevctx;
                                        this.pushNewRecursionContext(localctx, _startState, CashScriptParser.RULE_expression);
                                        this.state = 314;
                                        if (!(this.precpred(this._ctx, 10))) {
                                            throw this.createFailedPredicateException("this.precpred(this._ctx, 10)");
                                        }
                                        this.state = 315;
                                        localctx._op = this._input.LT(1);
                                        _la = this._input.LA(1);
                                        if (!(_la === 46 || _la === 47)) {
                                            localctx._op = this._errHandler.recoverInline(this);
                                        }
                                        else {
                                            this._errHandler.reportMatch(this);
                                            this.consume();
                                        }
                                        this.state = 316;
                                        localctx._right = this.expression(11);
                                    }
                                    break;
                                case 5:
                                    {
                                        localctx = new BinaryOpContext(this, new ExpressionContext(this, _parentctx, _parentState));
                                        localctx._left = _prevctx;
                                        this.pushNewRecursionContext(localctx, _startState, CashScriptParser.RULE_expression);
                                        this.state = 317;
                                        if (!(this.precpred(this._ctx, 9))) {
                                            throw this.createFailedPredicateException("this.precpred(this._ctx, 9)");
                                        }
                                        this.state = 318;
                                        localctx._op = this.match(CashScriptParser.T__47);
                                        this.state = 319;
                                        localctx._right = this.expression(10);
                                    }
                                    break;
                                case 6:
                                    {
                                        localctx = new BinaryOpContext(this, new ExpressionContext(this, _parentctx, _parentState));
                                        localctx._left = _prevctx;
                                        this.pushNewRecursionContext(localctx, _startState, CashScriptParser.RULE_expression);
                                        this.state = 320;
                                        if (!(this.precpred(this._ctx, 8))) {
                                            throw this.createFailedPredicateException("this.precpred(this._ctx, 8)");
                                        }
                                        this.state = 321;
                                        localctx._op = this.match(CashScriptParser.T__3);
                                        this.state = 322;
                                        localctx._right = this.expression(9);
                                    }
                                    break;
                                case 7:
                                    {
                                        localctx = new BinaryOpContext(this, new ExpressionContext(this, _parentctx, _parentState));
                                        localctx._left = _prevctx;
                                        this.pushNewRecursionContext(localctx, _startState, CashScriptParser.RULE_expression);
                                        this.state = 323;
                                        if (!(this.precpred(this._ctx, 7))) {
                                            throw this.createFailedPredicateException("this.precpred(this._ctx, 7)");
                                        }
                                        this.state = 324;
                                        localctx._op = this.match(CashScriptParser.T__48);
                                        this.state = 325;
                                        localctx._right = this.expression(8);
                                    }
                                    break;
                                case 8:
                                    {
                                        localctx = new BinaryOpContext(this, new ExpressionContext(this, _parentctx, _parentState));
                                        localctx._left = _prevctx;
                                        this.pushNewRecursionContext(localctx, _startState, CashScriptParser.RULE_expression);
                                        this.state = 326;
                                        if (!(this.precpred(this._ctx, 6))) {
                                            throw this.createFailedPredicateException("this.precpred(this._ctx, 6)");
                                        }
                                        this.state = 327;
                                        localctx._op = this.match(CashScriptParser.T__49);
                                        this.state = 328;
                                        localctx._right = this.expression(7);
                                    }
                                    break;
                                case 9:
                                    {
                                        localctx = new BinaryOpContext(this, new ExpressionContext(this, _parentctx, _parentState));
                                        localctx._left = _prevctx;
                                        this.pushNewRecursionContext(localctx, _startState, CashScriptParser.RULE_expression);
                                        this.state = 329;
                                        if (!(this.precpred(this._ctx, 5))) {
                                            throw this.createFailedPredicateException("this.precpred(this._ctx, 5)");
                                        }
                                        this.state = 330;
                                        localctx._op = this.match(CashScriptParser.T__50);
                                        this.state = 331;
                                        localctx._right = this.expression(6);
                                    }
                                    break;
                                case 10:
                                    {
                                        localctx = new TupleIndexOpContext(this, new ExpressionContext(this, _parentctx, _parentState));
                                        this.pushNewRecursionContext(localctx, _startState, CashScriptParser.RULE_expression);
                                        this.state = 332;
                                        if (!(this.precpred(this._ctx, 20))) {
                                            throw this.createFailedPredicateException("this.precpred(this._ctx, 20)");
                                        }
                                        this.state = 333;
                                        this.match(CashScriptParser.T__22);
                                        this.state = 334;
                                        localctx._index = this.match(CashScriptParser.NumberLiteral);
                                        this.state = 335;
                                        this.match(CashScriptParser.T__23);
                                    }
                                    break;
                                case 11:
                                    {
                                        localctx = new UnaryOpContext(this, new ExpressionContext(this, _parentctx, _parentState));
                                        this.pushNewRecursionContext(localctx, _startState, CashScriptParser.RULE_expression);
                                        this.state = 336;
                                        if (!(this.precpred(this._ctx, 17))) {
                                            throw this.createFailedPredicateException("this.precpred(this._ctx, 17)");
                                        }
                                        this.state = 337;
                                        localctx._op = this._input.LT(1);
                                        _la = this._input.LA(1);
                                        if (!(_la === 36 || _la === 37)) {
                                            localctx._op = this._errHandler.recoverInline(this);
                                        }
                                        else {
                                            this._errHandler.reportMatch(this);
                                            this.consume();
                                        }
                                    }
                                    break;
                                case 12:
                                    {
                                        localctx = new BinaryOpContext(this, new ExpressionContext(this, _parentctx, _parentState));
                                        localctx._left = _prevctx;
                                        this.pushNewRecursionContext(localctx, _startState, CashScriptParser.RULE_expression);
                                        this.state = 338;
                                        if (!(this.precpred(this._ctx, 16))) {
                                            throw this.createFailedPredicateException("this.precpred(this._ctx, 16)");
                                        }
                                        this.state = 339;
                                        localctx._op = this.match(CashScriptParser.T__37);
                                        this.state = 340;
                                        this.match(CashScriptParser.T__14);
                                        this.state = 341;
                                        localctx._right = this.expression(0);
                                        this.state = 342;
                                        this.match(CashScriptParser.T__16);
                                    }
                                    break;
                                case 13:
                                    {
                                        localctx = new SliceContext(this, new ExpressionContext(this, _parentctx, _parentState));
                                        localctx._element = _prevctx;
                                        this.pushNewRecursionContext(localctx, _startState, CashScriptParser.RULE_expression);
                                        this.state = 344;
                                        if (!(this.precpred(this._ctx, 15))) {
                                            throw this.createFailedPredicateException("this.precpred(this._ctx, 15)");
                                        }
                                        this.state = 345;
                                        this.match(CashScriptParser.T__38);
                                        this.state = 346;
                                        this.match(CashScriptParser.T__14);
                                        this.state = 347;
                                        localctx._start = this.expression(0);
                                        this.state = 348;
                                        this.match(CashScriptParser.T__15);
                                        this.state = 349;
                                        localctx._end = this.expression(0);
                                        this.state = 350;
                                        this.match(CashScriptParser.T__16);
                                    }
                                    break;
                            }
                        }
                    }
                    this.state = 356;
                    this._errHandler.sync(this);
                    _alt = this._interp.adaptivePredict(this._input, 29, this._ctx);
                }
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.unrollRecursionContexts(_parentctx);
        }
        return localctx;
    }
    // @RuleVersion(0)
    modifier() {
        let localctx = new ModifierContext(this, this._ctx, this.state);
        this.enterRule(localctx, 50, CashScriptParser.RULE_modifier);
        try {
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 357;
                this.match(CashScriptParser.T__51);
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    literal() {
        let localctx = new LiteralContext(this, this._ctx, this.state);
        this.enterRule(localctx, 52, CashScriptParser.RULE_literal);
        try {
            this.state = 364;
            this._errHandler.sync(this);
            switch (this._input.LA(1)) {
                case 60:
                    this.enterOuterAlt(localctx, 1);
                    {
                        this.state = 359;
                        this.match(CashScriptParser.BooleanLiteral);
                    }
                    break;
                case 62:
                    this.enterOuterAlt(localctx, 2);
                    {
                        this.state = 360;
                        this.numberLiteral();
                    }
                    break;
                case 67:
                    this.enterOuterAlt(localctx, 3);
                    {
                        this.state = 361;
                        this.match(CashScriptParser.StringLiteral);
                    }
                    break;
                case 68:
                    this.enterOuterAlt(localctx, 4);
                    {
                        this.state = 362;
                        this.match(CashScriptParser.DateLiteral);
                    }
                    break;
                case 69:
                    this.enterOuterAlt(localctx, 5);
                    {
                        this.state = 363;
                        this.match(CashScriptParser.HexLiteral);
                    }
                    break;
                default:
                    throw new NoViableAltException(this);
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    numberLiteral() {
        let localctx = new NumberLiteralContext(this, this._ctx, this.state);
        this.enterRule(localctx, 54, CashScriptParser.RULE_numberLiteral);
        try {
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 366;
                this.match(CashScriptParser.NumberLiteral);
                this.state = 368;
                this._errHandler.sync(this);
                switch (this._interp.adaptivePredict(this._input, 31, this._ctx)) {
                    case 1:
                        {
                            this.state = 367;
                            this.match(CashScriptParser.NumberUnit);
                        }
                        break;
                }
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    // @RuleVersion(0)
    typeName() {
        let localctx = new TypeNameContext(this, this._ctx, this.state);
        this.enterRule(localctx, 56, CashScriptParser.RULE_typeName);
        let _la;
        try {
            this.enterOuterAlt(localctx, 1);
            {
                this.state = 370;
                _la = this._input.LA(1);
                if (!(((((_la - 53)) & ~0x1F) === 0 && ((1 << (_la - 53)) & 4159) !== 0))) {
                    this._errHandler.recoverInline(this);
                }
                else {
                    this._errHandler.reportMatch(this);
                    this.consume();
                }
            }
        }
        catch (re) {
            if (re instanceof RecognitionException) {
                localctx.exception = re;
                this._errHandler.reportError(this, re);
                this._errHandler.recover(this, re);
            }
            else {
                throw re;
            }
        }
        finally {
            this.exitRule();
        }
        return localctx;
    }
    sempred(localctx, ruleIndex, predIndex) {
        switch (ruleIndex) {
            case 24:
                return this.expression_sempred(localctx, predIndex);
        }
        return true;
    }
    expression_sempred(localctx, predIndex) {
        switch (predIndex) {
            case 0:
                return this.precpred(this._ctx, 13);
            case 1:
                return this.precpred(this._ctx, 12);
            case 2:
                return this.precpred(this._ctx, 11);
            case 3:
                return this.precpred(this._ctx, 10);
            case 4:
                return this.precpred(this._ctx, 9);
            case 5:
                return this.precpred(this._ctx, 8);
            case 6:
                return this.precpred(this._ctx, 7);
            case 7:
                return this.precpred(this._ctx, 6);
            case 8:
                return this.precpred(this._ctx, 5);
            case 9:
                return this.precpred(this._ctx, 20);
            case 10:
                return this.precpred(this._ctx, 17);
            case 11:
                return this.precpred(this._ctx, 16);
            case 12:
                return this.precpred(this._ctx, 15);
        }
        return true;
    }
    static get _ATN() {
        if (!CashScriptParser.__ATN) {
            CashScriptParser.__ATN = new ATNDeserializer().deserialize(CashScriptParser._serializedATN);
        }
        return CashScriptParser.__ATN;
    }
}
CashScriptParser.T__0 = 1;
CashScriptParser.T__1 = 2;
CashScriptParser.T__2 = 3;
CashScriptParser.T__3 = 4;
CashScriptParser.T__4 = 5;
CashScriptParser.T__5 = 6;
CashScriptParser.T__6 = 7;
CashScriptParser.T__7 = 8;
CashScriptParser.T__8 = 9;
CashScriptParser.T__9 = 10;
CashScriptParser.T__10 = 11;
CashScriptParser.T__11 = 12;
CashScriptParser.T__12 = 13;
CashScriptParser.T__13 = 14;
CashScriptParser.T__14 = 15;
CashScriptParser.T__15 = 16;
CashScriptParser.T__16 = 17;
CashScriptParser.T__17 = 18;
CashScriptParser.T__18 = 19;
CashScriptParser.T__19 = 20;
CashScriptParser.T__20 = 21;
CashScriptParser.T__21 = 22;
CashScriptParser.T__22 = 23;
CashScriptParser.T__23 = 24;
CashScriptParser.T__24 = 25;
CashScriptParser.T__25 = 26;
CashScriptParser.T__26 = 27;
CashScriptParser.T__27 = 28;
CashScriptParser.T__28 = 29;
CashScriptParser.T__29 = 30;
CashScriptParser.T__30 = 31;
CashScriptParser.T__31 = 32;
CashScriptParser.T__32 = 33;
CashScriptParser.T__33 = 34;
CashScriptParser.T__34 = 35;
CashScriptParser.T__35 = 36;
CashScriptParser.T__36 = 37;
CashScriptParser.T__37 = 38;
CashScriptParser.T__38 = 39;
CashScriptParser.T__39 = 40;
CashScriptParser.T__40 = 41;
CashScriptParser.T__41 = 42;
CashScriptParser.T__42 = 43;
CashScriptParser.T__43 = 44;
CashScriptParser.T__44 = 45;
CashScriptParser.T__45 = 46;
CashScriptParser.T__46 = 47;
CashScriptParser.T__47 = 48;
CashScriptParser.T__48 = 49;
CashScriptParser.T__49 = 50;
CashScriptParser.T__50 = 51;
CashScriptParser.T__51 = 52;
CashScriptParser.T__52 = 53;
CashScriptParser.T__53 = 54;
CashScriptParser.T__54 = 55;
CashScriptParser.T__55 = 56;
CashScriptParser.T__56 = 57;
CashScriptParser.T__57 = 58;
CashScriptParser.VersionLiteral = 59;
CashScriptParser.BooleanLiteral = 60;
CashScriptParser.NumberUnit = 61;
CashScriptParser.NumberLiteral = 62;
CashScriptParser.NumberPart = 63;
CashScriptParser.ExponentPart = 64;
CashScriptParser.Bytes = 65;
CashScriptParser.Bound = 66;
CashScriptParser.StringLiteral = 67;
CashScriptParser.DateLiteral = 68;
CashScriptParser.HexLiteral = 69;
CashScriptParser.TxVar = 70;
CashScriptParser.NullaryOp = 71;
CashScriptParser.Identifier = 72;
CashScriptParser.WHITESPACE = 73;
CashScriptParser.COMMENT = 74;
CashScriptParser.LINE_COMMENT = 75;
CashScriptParser.EOF = Token.EOF;
CashScriptParser.RULE_sourceFile = 0;
CashScriptParser.RULE_pragmaDirective = 1;
CashScriptParser.RULE_pragmaName = 2;
CashScriptParser.RULE_pragmaValue = 3;
CashScriptParser.RULE_versionConstraint = 4;
CashScriptParser.RULE_versionOperator = 5;
CashScriptParser.RULE_contractDefinition = 6;
CashScriptParser.RULE_functionDefinition = 7;
CashScriptParser.RULE_parameterList = 8;
CashScriptParser.RULE_parameter = 9;
CashScriptParser.RULE_block = 10;
CashScriptParser.RULE_statement = 11;
CashScriptParser.RULE_variableDefinition = 12;
CashScriptParser.RULE_tupleAssignment = 13;
CashScriptParser.RULE_assignStatement = 14;
CashScriptParser.RULE_timeOpStatement = 15;
CashScriptParser.RULE_requireStatement = 16;
CashScriptParser.RULE_ifStatement = 17;
CashScriptParser.RULE_consoleStatement = 18;
CashScriptParser.RULE_requireMessage = 19;
CashScriptParser.RULE_consoleParameter = 20;
CashScriptParser.RULE_consoleParameterList = 21;
CashScriptParser.RULE_functionCall = 22;
CashScriptParser.RULE_expressionList = 23;
CashScriptParser.RULE_expression = 24;
CashScriptParser.RULE_modifier = 25;
CashScriptParser.RULE_literal = 26;
CashScriptParser.RULE_numberLiteral = 27;
CashScriptParser.RULE_typeName = 28;
CashScriptParser.literalNames = [null, "'pragma'",
    "';'", "'cashscript'",
    "'^'", "'~'",
    "'>='", "'>'",
    "'<'", "'<='",
    "'='", "'contract'",
    "'{'", "'}'",
    "'function'",
    "'('", "','",
    "')'", "'require'",
    "'if'", "'else'",
    "'console.log'",
    "'new'", "'['",
    "']'", "'tx.outputs'",
    "'.value'",
    "'.lockingBytecode'",
    "'.tokenCategory'",
    "'.nftCommitment'",
    "'.tokenAmount'",
    "'tx.inputs'",
    "'.outpointTransactionHash'",
    "'.outpointIndex'",
    "'.unlockingBytecode'",
    "'.sequenceNumber'",
    "'.reverse()'",
    "'.length'",
    "'.split'",
    "'.slice'",
    "'!'", "'-'",
    "'*'", "'/'",
    "'%'", "'+'",
    "'=='", "'!='",
    "'&'", "'|'",
    "'&&'", "'||'",
    "'constant'",
    "'int'", "'bool'",
    "'string'",
    "'pubkey'",
    "'sig'", "'datasig'"];
CashScriptParser.symbolicNames = [null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, null,
    null, "VersionLiteral",
    "BooleanLiteral",
    "NumberUnit",
    "NumberLiteral",
    "NumberPart",
    "ExponentPart",
    "Bytes", "Bound",
    "StringLiteral",
    "DateLiteral",
    "HexLiteral",
    "TxVar", "NullaryOp",
    "Identifier",
    "WHITESPACE",
    "COMMENT",
    "LINE_COMMENT"];
// tslint:disable:no-trailing-whitespace
CashScriptParser.ruleNames = [
    "sourceFile", "pragmaDirective", "pragmaName", "pragmaValue", "versionConstraint",
    "versionOperator", "contractDefinition", "functionDefinition", "parameterList",
    "parameter", "block", "statement", "variableDefinition", "tupleAssignment",
    "assignStatement", "timeOpStatement", "requireStatement", "ifStatement",
    "consoleStatement", "requireMessage", "consoleParameter", "consoleParameterList",
    "functionCall", "expressionList", "expression", "modifier", "literal",
    "numberLiteral", "typeName",
];
CashScriptParser._serializedATN = [4, 1, 75, 373, 2, 0, 7, 0, 2,
    1, 7, 1, 2, 2, 7, 2, 2, 3, 7, 3, 2, 4, 7, 4, 2, 5, 7, 5, 2, 6, 7, 6, 2, 7, 7, 7, 2, 8, 7, 8, 2, 9, 7, 9, 2,
    10, 7, 10, 2, 11, 7, 11, 2, 12, 7, 12, 2, 13, 7, 13, 2, 14, 7, 14, 2, 15, 7, 15, 2, 16, 7, 16, 2, 17,
    7, 17, 2, 18, 7, 18, 2, 19, 7, 19, 2, 20, 7, 20, 2, 21, 7, 21, 2, 22, 7, 22, 2, 23, 7, 23, 2, 24, 7,
    24, 2, 25, 7, 25, 2, 26, 7, 26, 2, 27, 7, 27, 2, 28, 7, 28, 1, 0, 5, 0, 60, 8, 0, 10, 0, 12, 0, 63,
    9, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 2, 1, 3, 1, 3, 3, 3, 77, 8, 3, 1, 4, 3, 4,
    80, 8, 4, 1, 4, 1, 4, 1, 5, 1, 5, 1, 6, 1, 6, 1, 6, 1, 6, 1, 6, 5, 6, 91, 8, 6, 10, 6, 12, 6, 94, 9, 6,
    1, 6, 1, 6, 1, 7, 1, 7, 1, 7, 1, 7, 1, 7, 5, 7, 103, 8, 7, 10, 7, 12, 7, 106, 9, 7, 1, 7, 1, 7, 1, 8, 1,
    8, 1, 8, 1, 8, 5, 8, 114, 8, 8, 10, 8, 12, 8, 117, 9, 8, 1, 8, 3, 8, 120, 8, 8, 3, 8, 122, 8, 8, 1, 8,
    1, 8, 1, 9, 1, 9, 1, 9, 1, 10, 1, 10, 5, 10, 131, 8, 10, 10, 10, 12, 10, 134, 9, 10, 1, 10, 1, 10,
    3, 10, 138, 8, 10, 1, 11, 1, 11, 1, 11, 1, 11, 1, 11, 1, 11, 1, 11, 3, 11, 147, 8, 11, 1, 12, 1, 12,
    5, 12, 151, 8, 12, 10, 12, 12, 12, 154, 9, 12, 1, 12, 1, 12, 1, 12, 1, 12, 1, 12, 1, 13, 1, 13, 1,
    13, 1, 13, 1, 13, 1, 13, 1, 13, 1, 13, 1, 13, 1, 14, 1, 14, 1, 14, 1, 14, 1, 14, 1, 15, 1, 15, 1, 15,
    1, 15, 1, 15, 1, 15, 1, 15, 3, 15, 182, 8, 15, 1, 15, 1, 15, 1, 15, 1, 16, 1, 16, 1, 16, 1, 16, 1,
    16, 3, 16, 192, 8, 16, 1, 16, 1, 16, 1, 16, 1, 17, 1, 17, 1, 17, 1, 17, 1, 17, 1, 17, 1, 17, 3, 17,
    204, 8, 17, 1, 18, 1, 18, 1, 18, 1, 18, 1, 19, 1, 19, 1, 20, 1, 20, 3, 20, 214, 8, 20, 1, 21, 1, 21,
    1, 21, 1, 21, 5, 21, 220, 8, 21, 10, 21, 12, 21, 223, 9, 21, 1, 21, 3, 21, 226, 8, 21, 3, 21, 228,
    8, 21, 1, 21, 1, 21, 1, 22, 1, 22, 1, 22, 1, 23, 1, 23, 1, 23, 1, 23, 5, 23, 239, 8, 23, 10, 23, 12,
    23, 242, 9, 23, 1, 23, 3, 23, 245, 8, 23, 3, 23, 247, 8, 23, 1, 23, 1, 23, 1, 24, 1, 24, 1, 24, 1,
    24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 3, 24, 261, 8, 24, 1, 24, 3, 24, 264, 8, 24, 1, 24,
    1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1,
    24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 5, 24, 290, 8, 24, 10, 24, 12, 24, 293,
    9, 24, 1, 24, 3, 24, 296, 8, 24, 3, 24, 298, 8, 24, 1, 24, 1, 24, 1, 24, 1, 24, 3, 24, 304, 8, 24,
    1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1,
    24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24,
    1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1, 24, 1,
    24, 1, 24, 1, 24, 1, 24, 5, 24, 353, 8, 24, 10, 24, 12, 24, 356, 9, 24, 1, 25, 1, 25, 1, 26, 1, 26,
    1, 26, 1, 26, 1, 26, 3, 26, 365, 8, 26, 1, 27, 1, 27, 3, 27, 369, 8, 27, 1, 28, 1, 28, 1, 28, 0, 1,
    48, 29, 0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46,
    48, 50, 52, 54, 56, 0, 10, 1, 0, 4, 10, 1, 0, 26, 30, 2, 0, 26, 30, 32, 35, 1, 0, 40, 41, 1, 0, 42,
    44, 2, 0, 41, 41, 45, 45, 1, 0, 6, 9, 1, 0, 46, 47, 1, 0, 36, 37, 2, 0, 53, 58, 65, 65, 403, 0, 61,
    1, 0, 0, 0, 2, 67, 1, 0, 0, 0, 4, 72, 1, 0, 0, 0, 6, 74, 1, 0, 0, 0, 8, 79, 1, 0, 0, 0, 10, 83, 1, 0, 0,
    0, 12, 85, 1, 0, 0, 0, 14, 97, 1, 0, 0, 0, 16, 109, 1, 0, 0, 0, 18, 125, 1, 0, 0, 0, 20, 137, 1, 0,
    0, 0, 22, 146, 1, 0, 0, 0, 24, 148, 1, 0, 0, 0, 26, 160, 1, 0, 0, 0, 28, 169, 1, 0, 0, 0, 30, 174,
    1, 0, 0, 0, 32, 186, 1, 0, 0, 0, 34, 196, 1, 0, 0, 0, 36, 205, 1, 0, 0, 0, 38, 209, 1, 0, 0, 0, 40,
    213, 1, 0, 0, 0, 42, 215, 1, 0, 0, 0, 44, 231, 1, 0, 0, 0, 46, 234, 1, 0, 0, 0, 48, 303, 1, 0, 0, 0,
    50, 357, 1, 0, 0, 0, 52, 364, 1, 0, 0, 0, 54, 366, 1, 0, 0, 0, 56, 370, 1, 0, 0, 0, 58, 60, 3, 2, 1,
    0, 59, 58, 1, 0, 0, 0, 60, 63, 1, 0, 0, 0, 61, 59, 1, 0, 0, 0, 61, 62, 1, 0, 0, 0, 62, 64, 1, 0, 0, 0,
    63, 61, 1, 0, 0, 0, 64, 65, 3, 12, 6, 0, 65, 66, 5, 0, 0, 1, 66, 1, 1, 0, 0, 0, 67, 68, 5, 1, 0, 0, 68,
    69, 3, 4, 2, 0, 69, 70, 3, 6, 3, 0, 70, 71, 5, 2, 0, 0, 71, 3, 1, 0, 0, 0, 72, 73, 5, 3, 0, 0, 73, 5,
    1, 0, 0, 0, 74, 76, 3, 8, 4, 0, 75, 77, 3, 8, 4, 0, 76, 75, 1, 0, 0, 0, 76, 77, 1, 0, 0, 0, 77, 7, 1,
    0, 0, 0, 78, 80, 3, 10, 5, 0, 79, 78, 1, 0, 0, 0, 79, 80, 1, 0, 0, 0, 80, 81, 1, 0, 0, 0, 81, 82, 5,
    59, 0, 0, 82, 9, 1, 0, 0, 0, 83, 84, 7, 0, 0, 0, 84, 11, 1, 0, 0, 0, 85, 86, 5, 11, 0, 0, 86, 87, 5,
    72, 0, 0, 87, 88, 3, 16, 8, 0, 88, 92, 5, 12, 0, 0, 89, 91, 3, 14, 7, 0, 90, 89, 1, 0, 0, 0, 91, 94,
    1, 0, 0, 0, 92, 90, 1, 0, 0, 0, 92, 93, 1, 0, 0, 0, 93, 95, 1, 0, 0, 0, 94, 92, 1, 0, 0, 0, 95, 96, 5,
    13, 0, 0, 96, 13, 1, 0, 0, 0, 97, 98, 5, 14, 0, 0, 98, 99, 5, 72, 0, 0, 99, 100, 3, 16, 8, 0, 100,
    104, 5, 12, 0, 0, 101, 103, 3, 22, 11, 0, 102, 101, 1, 0, 0, 0, 103, 106, 1, 0, 0, 0, 104, 102,
    1, 0, 0, 0, 104, 105, 1, 0, 0, 0, 105, 107, 1, 0, 0, 0, 106, 104, 1, 0, 0, 0, 107, 108, 5, 13, 0,
    0, 108, 15, 1, 0, 0, 0, 109, 121, 5, 15, 0, 0, 110, 115, 3, 18, 9, 0, 111, 112, 5, 16, 0, 0, 112,
    114, 3, 18, 9, 0, 113, 111, 1, 0, 0, 0, 114, 117, 1, 0, 0, 0, 115, 113, 1, 0, 0, 0, 115, 116, 1,
    0, 0, 0, 116, 119, 1, 0, 0, 0, 117, 115, 1, 0, 0, 0, 118, 120, 5, 16, 0, 0, 119, 118, 1, 0, 0, 0,
    119, 120, 1, 0, 0, 0, 120, 122, 1, 0, 0, 0, 121, 110, 1, 0, 0, 0, 121, 122, 1, 0, 0, 0, 122, 123,
    1, 0, 0, 0, 123, 124, 5, 17, 0, 0, 124, 17, 1, 0, 0, 0, 125, 126, 3, 56, 28, 0, 126, 127, 5, 72,
    0, 0, 127, 19, 1, 0, 0, 0, 128, 132, 5, 12, 0, 0, 129, 131, 3, 22, 11, 0, 130, 129, 1, 0, 0, 0, 131,
    134, 1, 0, 0, 0, 132, 130, 1, 0, 0, 0, 132, 133, 1, 0, 0, 0, 133, 135, 1, 0, 0, 0, 134, 132, 1, 0,
    0, 0, 135, 138, 5, 13, 0, 0, 136, 138, 3, 22, 11, 0, 137, 128, 1, 0, 0, 0, 137, 136, 1, 0, 0, 0,
    138, 21, 1, 0, 0, 0, 139, 147, 3, 24, 12, 0, 140, 147, 3, 26, 13, 0, 141, 147, 3, 28, 14, 0, 142,
    147, 3, 30, 15, 0, 143, 147, 3, 32, 16, 0, 144, 147, 3, 34, 17, 0, 145, 147, 3, 36, 18, 0, 146,
    139, 1, 0, 0, 0, 146, 140, 1, 0, 0, 0, 146, 141, 1, 0, 0, 0, 146, 142, 1, 0, 0, 0, 146, 143, 1, 0,
    0, 0, 146, 144, 1, 0, 0, 0, 146, 145, 1, 0, 0, 0, 147, 23, 1, 0, 0, 0, 148, 152, 3, 56, 28, 0, 149,
    151, 3, 50, 25, 0, 150, 149, 1, 0, 0, 0, 151, 154, 1, 0, 0, 0, 152, 150, 1, 0, 0, 0, 152, 153, 1,
    0, 0, 0, 153, 155, 1, 0, 0, 0, 154, 152, 1, 0, 0, 0, 155, 156, 5, 72, 0, 0, 156, 157, 5, 10, 0, 0,
    157, 158, 3, 48, 24, 0, 158, 159, 5, 2, 0, 0, 159, 25, 1, 0, 0, 0, 160, 161, 3, 56, 28, 0, 161,
    162, 5, 72, 0, 0, 162, 163, 5, 16, 0, 0, 163, 164, 3, 56, 28, 0, 164, 165, 5, 72, 0, 0, 165, 166,
    5, 10, 0, 0, 166, 167, 3, 48, 24, 0, 167, 168, 5, 2, 0, 0, 168, 27, 1, 0, 0, 0, 169, 170, 5, 72,
    0, 0, 170, 171, 5, 10, 0, 0, 171, 172, 3, 48, 24, 0, 172, 173, 5, 2, 0, 0, 173, 29, 1, 0, 0, 0, 174,
    175, 5, 18, 0, 0, 175, 176, 5, 15, 0, 0, 176, 177, 5, 70, 0, 0, 177, 178, 5, 6, 0, 0, 178, 181,
    3, 48, 24, 0, 179, 180, 5, 16, 0, 0, 180, 182, 3, 38, 19, 0, 181, 179, 1, 0, 0, 0, 181, 182, 1,
    0, 0, 0, 182, 183, 1, 0, 0, 0, 183, 184, 5, 17, 0, 0, 184, 185, 5, 2, 0, 0, 185, 31, 1, 0, 0, 0, 186,
    187, 5, 18, 0, 0, 187, 188, 5, 15, 0, 0, 188, 191, 3, 48, 24, 0, 189, 190, 5, 16, 0, 0, 190, 192,
    3, 38, 19, 0, 191, 189, 1, 0, 0, 0, 191, 192, 1, 0, 0, 0, 192, 193, 1, 0, 0, 0, 193, 194, 5, 17,
    0, 0, 194, 195, 5, 2, 0, 0, 195, 33, 1, 0, 0, 0, 196, 197, 5, 19, 0, 0, 197, 198, 5, 15, 0, 0, 198,
    199, 3, 48, 24, 0, 199, 200, 5, 17, 0, 0, 200, 203, 3, 20, 10, 0, 201, 202, 5, 20, 0, 0, 202, 204,
    3, 20, 10, 0, 203, 201, 1, 0, 0, 0, 203, 204, 1, 0, 0, 0, 204, 35, 1, 0, 0, 0, 205, 206, 5, 21, 0,
    0, 206, 207, 3, 42, 21, 0, 207, 208, 5, 2, 0, 0, 208, 37, 1, 0, 0, 0, 209, 210, 5, 67, 0, 0, 210,
    39, 1, 0, 0, 0, 211, 214, 5, 72, 0, 0, 212, 214, 3, 52, 26, 0, 213, 211, 1, 0, 0, 0, 213, 212, 1,
    0, 0, 0, 214, 41, 1, 0, 0, 0, 215, 227, 5, 15, 0, 0, 216, 221, 3, 40, 20, 0, 217, 218, 5, 16, 0,
    0, 218, 220, 3, 40, 20, 0, 219, 217, 1, 0, 0, 0, 220, 223, 1, 0, 0, 0, 221, 219, 1, 0, 0, 0, 221,
    222, 1, 0, 0, 0, 222, 225, 1, 0, 0, 0, 223, 221, 1, 0, 0, 0, 224, 226, 5, 16, 0, 0, 225, 224, 1,
    0, 0, 0, 225, 226, 1, 0, 0, 0, 226, 228, 1, 0, 0, 0, 227, 216, 1, 0, 0, 0, 227, 228, 1, 0, 0, 0, 228,
    229, 1, 0, 0, 0, 229, 230, 5, 17, 0, 0, 230, 43, 1, 0, 0, 0, 231, 232, 5, 72, 0, 0, 232, 233, 3,
    46, 23, 0, 233, 45, 1, 0, 0, 0, 234, 246, 5, 15, 0, 0, 235, 240, 3, 48, 24, 0, 236, 237, 5, 16,
    0, 0, 237, 239, 3, 48, 24, 0, 238, 236, 1, 0, 0, 0, 239, 242, 1, 0, 0, 0, 240, 238, 1, 0, 0, 0, 240,
    241, 1, 0, 0, 0, 241, 244, 1, 0, 0, 0, 242, 240, 1, 0, 0, 0, 243, 245, 5, 16, 0, 0, 244, 243, 1,
    0, 0, 0, 244, 245, 1, 0, 0, 0, 245, 247, 1, 0, 0, 0, 246, 235, 1, 0, 0, 0, 246, 247, 1, 0, 0, 0, 247,
    248, 1, 0, 0, 0, 248, 249, 5, 17, 0, 0, 249, 47, 1, 0, 0, 0, 250, 251, 6, 24, -1, 0, 251, 252, 5,
    15, 0, 0, 252, 253, 3, 48, 24, 0, 253, 254, 5, 17, 0, 0, 254, 304, 1, 0, 0, 0, 255, 256, 3, 56,
    28, 0, 256, 257, 5, 15, 0, 0, 257, 260, 3, 48, 24, 0, 258, 259, 5, 16, 0, 0, 259, 261, 3, 48, 24,
    0, 260, 258, 1, 0, 0, 0, 260, 261, 1, 0, 0, 0, 261, 263, 1, 0, 0, 0, 262, 264, 5, 16, 0, 0, 263,
    262, 1, 0, 0, 0, 263, 264, 1, 0, 0, 0, 264, 265, 1, 0, 0, 0, 265, 266, 5, 17, 0, 0, 266, 304, 1,
    0, 0, 0, 267, 304, 3, 44, 22, 0, 268, 269, 5, 22, 0, 0, 269, 270, 5, 72, 0, 0, 270, 304, 3, 46,
    23, 0, 271, 272, 5, 25, 0, 0, 272, 273, 5, 23, 0, 0, 273, 274, 3, 48, 24, 0, 274, 275, 5, 24, 0,
    0, 275, 276, 7, 1, 0, 0, 276, 304, 1, 0, 0, 0, 277, 278, 5, 31, 0, 0, 278, 279, 5, 23, 0, 0, 279,
    280, 3, 48, 24, 0, 280, 281, 5, 24, 0, 0, 281, 282, 7, 2, 0, 0, 282, 304, 1, 0, 0, 0, 283, 284,
    7, 3, 0, 0, 284, 304, 3, 48, 24, 14, 285, 297, 5, 23, 0, 0, 286, 291, 3, 48, 24, 0, 287, 288, 5,
    16, 0, 0, 288, 290, 3, 48, 24, 0, 289, 287, 1, 0, 0, 0, 290, 293, 1, 0, 0, 0, 291, 289, 1, 0, 0,
    0, 291, 292, 1, 0, 0, 0, 292, 295, 1, 0, 0, 0, 293, 291, 1, 0, 0, 0, 294, 296, 5, 16, 0, 0, 295,
    294, 1, 0, 0, 0, 295, 296, 1, 0, 0, 0, 296, 298, 1, 0, 0, 0, 297, 286, 1, 0, 0, 0, 297, 298, 1, 0,
    0, 0, 298, 299, 1, 0, 0, 0, 299, 304, 5, 24, 0, 0, 300, 304, 5, 71, 0, 0, 301, 304, 5, 72, 0, 0,
    302, 304, 3, 52, 26, 0, 303, 250, 1, 0, 0, 0, 303, 255, 1, 0, 0, 0, 303, 267, 1, 0, 0, 0, 303, 268,
    1, 0, 0, 0, 303, 271, 1, 0, 0, 0, 303, 277, 1, 0, 0, 0, 303, 283, 1, 0, 0, 0, 303, 285, 1, 0, 0, 0,
    303, 300, 1, 0, 0, 0, 303, 301, 1, 0, 0, 0, 303, 302, 1, 0, 0, 0, 304, 354, 1, 0, 0, 0, 305, 306,
    10, 13, 0, 0, 306, 307, 7, 4, 0, 0, 307, 353, 3, 48, 24, 14, 308, 309, 10, 12, 0, 0, 309, 310,
    7, 5, 0, 0, 310, 353, 3, 48, 24, 13, 311, 312, 10, 11, 0, 0, 312, 313, 7, 6, 0, 0, 313, 353, 3,
    48, 24, 12, 314, 315, 10, 10, 0, 0, 315, 316, 7, 7, 0, 0, 316, 353, 3, 48, 24, 11, 317, 318, 10,
    9, 0, 0, 318, 319, 5, 48, 0, 0, 319, 353, 3, 48, 24, 10, 320, 321, 10, 8, 0, 0, 321, 322, 5, 4,
    0, 0, 322, 353, 3, 48, 24, 9, 323, 324, 10, 7, 0, 0, 324, 325, 5, 49, 0, 0, 325, 353, 3, 48, 24,
    8, 326, 327, 10, 6, 0, 0, 327, 328, 5, 50, 0, 0, 328, 353, 3, 48, 24, 7, 329, 330, 10, 5, 0, 0,
    330, 331, 5, 51, 0, 0, 331, 353, 3, 48, 24, 6, 332, 333, 10, 20, 0, 0, 333, 334, 5, 23, 0, 0, 334,
    335, 5, 62, 0, 0, 335, 353, 5, 24, 0, 0, 336, 337, 10, 17, 0, 0, 337, 353, 7, 8, 0, 0, 338, 339,
    10, 16, 0, 0, 339, 340, 5, 38, 0, 0, 340, 341, 5, 15, 0, 0, 341, 342, 3, 48, 24, 0, 342, 343, 5,
    17, 0, 0, 343, 353, 1, 0, 0, 0, 344, 345, 10, 15, 0, 0, 345, 346, 5, 39, 0, 0, 346, 347, 5, 15,
    0, 0, 347, 348, 3, 48, 24, 0, 348, 349, 5, 16, 0, 0, 349, 350, 3, 48, 24, 0, 350, 351, 5, 17, 0,
    0, 351, 353, 1, 0, 0, 0, 352, 305, 1, 0, 0, 0, 352, 308, 1, 0, 0, 0, 352, 311, 1, 0, 0, 0, 352, 314,
    1, 0, 0, 0, 352, 317, 1, 0, 0, 0, 352, 320, 1, 0, 0, 0, 352, 323, 1, 0, 0, 0, 352, 326, 1, 0, 0, 0,
    352, 329, 1, 0, 0, 0, 352, 332, 1, 0, 0, 0, 352, 336, 1, 0, 0, 0, 352, 338, 1, 0, 0, 0, 352, 344,
    1, 0, 0, 0, 353, 356, 1, 0, 0, 0, 354, 352, 1, 0, 0, 0, 354, 355, 1, 0, 0, 0, 355, 49, 1, 0, 0, 0,
    356, 354, 1, 0, 0, 0, 357, 358, 5, 52, 0, 0, 358, 51, 1, 0, 0, 0, 359, 365, 5, 60, 0, 0, 360, 365,
    3, 54, 27, 0, 361, 365, 5, 67, 0, 0, 362, 365, 5, 68, 0, 0, 363, 365, 5, 69, 0, 0, 364, 359, 1,
    0, 0, 0, 364, 360, 1, 0, 0, 0, 364, 361, 1, 0, 0, 0, 364, 362, 1, 0, 0, 0, 364, 363, 1, 0, 0, 0, 365,
    53, 1, 0, 0, 0, 366, 368, 5, 62, 0, 0, 367, 369, 5, 61, 0, 0, 368, 367, 1, 0, 0, 0, 368, 369, 1,
    0, 0, 0, 369, 55, 1, 0, 0, 0, 370, 371, 7, 9, 0, 0, 371, 57, 1, 0, 0, 0, 32, 61, 76, 79, 92, 104,
    115, 119, 121, 132, 137, 146, 152, 181, 191, 203, 213, 221, 225, 227, 240, 244, 246, 260,
    263, 291, 295, 297, 303, 352, 354, 364, 368];
CashScriptParser.DecisionsToDFA = CashScriptParser._ATN.decisionToState.map((ds, index) => new DFA(ds, index));
export default CashScriptParser;
export class SourceFileContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    contractDefinition() {
        return this.getTypedRuleContext(ContractDefinitionContext, 0);
    }
    EOF() {
        return this.getToken(CashScriptParser.EOF, 0);
    }
    pragmaDirective_list() {
        return this.getTypedRuleContexts(PragmaDirectiveContext);
    }
    pragmaDirective(i) {
        return this.getTypedRuleContext(PragmaDirectiveContext, i);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_sourceFile;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitSourceFile) {
            return visitor.visitSourceFile(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class PragmaDirectiveContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    pragmaName() {
        return this.getTypedRuleContext(PragmaNameContext, 0);
    }
    pragmaValue() {
        return this.getTypedRuleContext(PragmaValueContext, 0);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_pragmaDirective;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitPragmaDirective) {
            return visitor.visitPragmaDirective(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class PragmaNameContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    get ruleIndex() {
        return CashScriptParser.RULE_pragmaName;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitPragmaName) {
            return visitor.visitPragmaName(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class PragmaValueContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    versionConstraint_list() {
        return this.getTypedRuleContexts(VersionConstraintContext);
    }
    versionConstraint(i) {
        return this.getTypedRuleContext(VersionConstraintContext, i);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_pragmaValue;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitPragmaValue) {
            return visitor.visitPragmaValue(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class VersionConstraintContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    VersionLiteral() {
        return this.getToken(CashScriptParser.VersionLiteral, 0);
    }
    versionOperator() {
        return this.getTypedRuleContext(VersionOperatorContext, 0);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_versionConstraint;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitVersionConstraint) {
            return visitor.visitVersionConstraint(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class VersionOperatorContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    get ruleIndex() {
        return CashScriptParser.RULE_versionOperator;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitVersionOperator) {
            return visitor.visitVersionOperator(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class ContractDefinitionContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    Identifier() {
        return this.getToken(CashScriptParser.Identifier, 0);
    }
    parameterList() {
        return this.getTypedRuleContext(ParameterListContext, 0);
    }
    functionDefinition_list() {
        return this.getTypedRuleContexts(FunctionDefinitionContext);
    }
    functionDefinition(i) {
        return this.getTypedRuleContext(FunctionDefinitionContext, i);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_contractDefinition;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitContractDefinition) {
            return visitor.visitContractDefinition(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class FunctionDefinitionContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    Identifier() {
        return this.getToken(CashScriptParser.Identifier, 0);
    }
    parameterList() {
        return this.getTypedRuleContext(ParameterListContext, 0);
    }
    statement_list() {
        return this.getTypedRuleContexts(StatementContext);
    }
    statement(i) {
        return this.getTypedRuleContext(StatementContext, i);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_functionDefinition;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitFunctionDefinition) {
            return visitor.visitFunctionDefinition(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class ParameterListContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    parameter_list() {
        return this.getTypedRuleContexts(ParameterContext);
    }
    parameter(i) {
        return this.getTypedRuleContext(ParameterContext, i);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_parameterList;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitParameterList) {
            return visitor.visitParameterList(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class ParameterContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    typeName() {
        return this.getTypedRuleContext(TypeNameContext, 0);
    }
    Identifier() {
        return this.getToken(CashScriptParser.Identifier, 0);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_parameter;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitParameter) {
            return visitor.visitParameter(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class BlockContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    statement_list() {
        return this.getTypedRuleContexts(StatementContext);
    }
    statement(i) {
        return this.getTypedRuleContext(StatementContext, i);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_block;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitBlock) {
            return visitor.visitBlock(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class StatementContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    variableDefinition() {
        return this.getTypedRuleContext(VariableDefinitionContext, 0);
    }
    tupleAssignment() {
        return this.getTypedRuleContext(TupleAssignmentContext, 0);
    }
    assignStatement() {
        return this.getTypedRuleContext(AssignStatementContext, 0);
    }
    timeOpStatement() {
        return this.getTypedRuleContext(TimeOpStatementContext, 0);
    }
    requireStatement() {
        return this.getTypedRuleContext(RequireStatementContext, 0);
    }
    ifStatement() {
        return this.getTypedRuleContext(IfStatementContext, 0);
    }
    consoleStatement() {
        return this.getTypedRuleContext(ConsoleStatementContext, 0);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_statement;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitStatement) {
            return visitor.visitStatement(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class VariableDefinitionContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    typeName() {
        return this.getTypedRuleContext(TypeNameContext, 0);
    }
    Identifier() {
        return this.getToken(CashScriptParser.Identifier, 0);
    }
    expression() {
        return this.getTypedRuleContext(ExpressionContext, 0);
    }
    modifier_list() {
        return this.getTypedRuleContexts(ModifierContext);
    }
    modifier(i) {
        return this.getTypedRuleContext(ModifierContext, i);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_variableDefinition;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitVariableDefinition) {
            return visitor.visitVariableDefinition(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class TupleAssignmentContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    typeName_list() {
        return this.getTypedRuleContexts(TypeNameContext);
    }
    typeName(i) {
        return this.getTypedRuleContext(TypeNameContext, i);
    }
    Identifier_list() {
        return this.getTokens(CashScriptParser.Identifier);
    }
    Identifier(i) {
        return this.getToken(CashScriptParser.Identifier, i);
    }
    expression() {
        return this.getTypedRuleContext(ExpressionContext, 0);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_tupleAssignment;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitTupleAssignment) {
            return visitor.visitTupleAssignment(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class AssignStatementContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    Identifier() {
        return this.getToken(CashScriptParser.Identifier, 0);
    }
    expression() {
        return this.getTypedRuleContext(ExpressionContext, 0);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_assignStatement;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitAssignStatement) {
            return visitor.visitAssignStatement(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class TimeOpStatementContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    TxVar() {
        return this.getToken(CashScriptParser.TxVar, 0);
    }
    expression() {
        return this.getTypedRuleContext(ExpressionContext, 0);
    }
    requireMessage() {
        return this.getTypedRuleContext(RequireMessageContext, 0);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_timeOpStatement;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitTimeOpStatement) {
            return visitor.visitTimeOpStatement(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class RequireStatementContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    expression() {
        return this.getTypedRuleContext(ExpressionContext, 0);
    }
    requireMessage() {
        return this.getTypedRuleContext(RequireMessageContext, 0);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_requireStatement;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitRequireStatement) {
            return visitor.visitRequireStatement(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class IfStatementContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    expression() {
        return this.getTypedRuleContext(ExpressionContext, 0);
    }
    block_list() {
        return this.getTypedRuleContexts(BlockContext);
    }
    block(i) {
        return this.getTypedRuleContext(BlockContext, i);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_ifStatement;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitIfStatement) {
            return visitor.visitIfStatement(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class ConsoleStatementContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    consoleParameterList() {
        return this.getTypedRuleContext(ConsoleParameterListContext, 0);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_consoleStatement;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitConsoleStatement) {
            return visitor.visitConsoleStatement(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class RequireMessageContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    StringLiteral() {
        return this.getToken(CashScriptParser.StringLiteral, 0);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_requireMessage;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitRequireMessage) {
            return visitor.visitRequireMessage(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class ConsoleParameterContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    Identifier() {
        return this.getToken(CashScriptParser.Identifier, 0);
    }
    literal() {
        return this.getTypedRuleContext(LiteralContext, 0);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_consoleParameter;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitConsoleParameter) {
            return visitor.visitConsoleParameter(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class ConsoleParameterListContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    consoleParameter_list() {
        return this.getTypedRuleContexts(ConsoleParameterContext);
    }
    consoleParameter(i) {
        return this.getTypedRuleContext(ConsoleParameterContext, i);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_consoleParameterList;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitConsoleParameterList) {
            return visitor.visitConsoleParameterList(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class FunctionCallContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    Identifier() {
        return this.getToken(CashScriptParser.Identifier, 0);
    }
    expressionList() {
        return this.getTypedRuleContext(ExpressionListContext, 0);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_functionCall;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitFunctionCall) {
            return visitor.visitFunctionCall(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class ExpressionListContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    expression_list() {
        return this.getTypedRuleContexts(ExpressionContext);
    }
    expression(i) {
        return this.getTypedRuleContext(ExpressionContext, i);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_expressionList;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitExpressionList) {
            return visitor.visitExpressionList(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class ExpressionContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    get ruleIndex() {
        return CashScriptParser.RULE_expression;
    }
    copyFrom(ctx) {
        super.copyFrom(ctx);
    }
}
export class CastContext extends ExpressionContext {
    constructor(parser, ctx) {
        super(parser, ctx.parentCtx, ctx.invokingState);
        super.copyFrom(ctx);
    }
    typeName() {
        return this.getTypedRuleContext(TypeNameContext, 0);
    }
    expression_list() {
        return this.getTypedRuleContexts(ExpressionContext);
    }
    expression(i) {
        return this.getTypedRuleContext(ExpressionContext, i);
    }
    // @Override
    accept(visitor) {
        if (visitor.visitCast) {
            return visitor.visitCast(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class UnaryIntrospectionOpContext extends ExpressionContext {
    constructor(parser, ctx) {
        super(parser, ctx.parentCtx, ctx.invokingState);
        super.copyFrom(ctx);
    }
    expression() {
        return this.getTypedRuleContext(ExpressionContext, 0);
    }
    // @Override
    accept(visitor) {
        if (visitor.visitUnaryIntrospectionOp) {
            return visitor.visitUnaryIntrospectionOp(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class UnaryOpContext extends ExpressionContext {
    constructor(parser, ctx) {
        super(parser, ctx.parentCtx, ctx.invokingState);
        super.copyFrom(ctx);
    }
    expression() {
        return this.getTypedRuleContext(ExpressionContext, 0);
    }
    // @Override
    accept(visitor) {
        if (visitor.visitUnaryOp) {
            return visitor.visitUnaryOp(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class LiteralExpressionContext extends ExpressionContext {
    constructor(parser, ctx) {
        super(parser, ctx.parentCtx, ctx.invokingState);
        super.copyFrom(ctx);
    }
    literal() {
        return this.getTypedRuleContext(LiteralContext, 0);
    }
    // @Override
    accept(visitor) {
        if (visitor.visitLiteralExpression) {
            return visitor.visitLiteralExpression(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class FunctionCallExpressionContext extends ExpressionContext {
    constructor(parser, ctx) {
        super(parser, ctx.parentCtx, ctx.invokingState);
        super.copyFrom(ctx);
    }
    functionCall() {
        return this.getTypedRuleContext(FunctionCallContext, 0);
    }
    // @Override
    accept(visitor) {
        if (visitor.visitFunctionCallExpression) {
            return visitor.visitFunctionCallExpression(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class ArrayContext extends ExpressionContext {
    constructor(parser, ctx) {
        super(parser, ctx.parentCtx, ctx.invokingState);
        super.copyFrom(ctx);
    }
    expression_list() {
        return this.getTypedRuleContexts(ExpressionContext);
    }
    expression(i) {
        return this.getTypedRuleContext(ExpressionContext, i);
    }
    // @Override
    accept(visitor) {
        if (visitor.visitArray) {
            return visitor.visitArray(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class IdentifierContext extends ExpressionContext {
    constructor(parser, ctx) {
        super(parser, ctx.parentCtx, ctx.invokingState);
        super.copyFrom(ctx);
    }
    Identifier() {
        return this.getToken(CashScriptParser.Identifier, 0);
    }
    // @Override
    accept(visitor) {
        if (visitor.visitIdentifier) {
            return visitor.visitIdentifier(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class SliceContext extends ExpressionContext {
    constructor(parser, ctx) {
        super(parser, ctx.parentCtx, ctx.invokingState);
        super.copyFrom(ctx);
    }
    expression_list() {
        return this.getTypedRuleContexts(ExpressionContext);
    }
    expression(i) {
        return this.getTypedRuleContext(ExpressionContext, i);
    }
    // @Override
    accept(visitor) {
        if (visitor.visitSlice) {
            return visitor.visitSlice(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class TupleIndexOpContext extends ExpressionContext {
    constructor(parser, ctx) {
        super(parser, ctx.parentCtx, ctx.invokingState);
        super.copyFrom(ctx);
    }
    expression() {
        return this.getTypedRuleContext(ExpressionContext, 0);
    }
    NumberLiteral() {
        return this.getToken(CashScriptParser.NumberLiteral, 0);
    }
    // @Override
    accept(visitor) {
        if (visitor.visitTupleIndexOp) {
            return visitor.visitTupleIndexOp(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class InstantiationContext extends ExpressionContext {
    constructor(parser, ctx) {
        super(parser, ctx.parentCtx, ctx.invokingState);
        super.copyFrom(ctx);
    }
    Identifier() {
        return this.getToken(CashScriptParser.Identifier, 0);
    }
    expressionList() {
        return this.getTypedRuleContext(ExpressionListContext, 0);
    }
    // @Override
    accept(visitor) {
        if (visitor.visitInstantiation) {
            return visitor.visitInstantiation(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class NullaryOpContext extends ExpressionContext {
    constructor(parser, ctx) {
        super(parser, ctx.parentCtx, ctx.invokingState);
        super.copyFrom(ctx);
    }
    NullaryOp() {
        return this.getToken(CashScriptParser.NullaryOp, 0);
    }
    // @Override
    accept(visitor) {
        if (visitor.visitNullaryOp) {
            return visitor.visitNullaryOp(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class ParenthesisedContext extends ExpressionContext {
    constructor(parser, ctx) {
        super(parser, ctx.parentCtx, ctx.invokingState);
        super.copyFrom(ctx);
    }
    expression() {
        return this.getTypedRuleContext(ExpressionContext, 0);
    }
    // @Override
    accept(visitor) {
        if (visitor.visitParenthesised) {
            return visitor.visitParenthesised(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class BinaryOpContext extends ExpressionContext {
    constructor(parser, ctx) {
        super(parser, ctx.parentCtx, ctx.invokingState);
        super.copyFrom(ctx);
    }
    expression_list() {
        return this.getTypedRuleContexts(ExpressionContext);
    }
    expression(i) {
        return this.getTypedRuleContext(ExpressionContext, i);
    }
    // @Override
    accept(visitor) {
        if (visitor.visitBinaryOp) {
            return visitor.visitBinaryOp(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class ModifierContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    get ruleIndex() {
        return CashScriptParser.RULE_modifier;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitModifier) {
            return visitor.visitModifier(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class LiteralContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    BooleanLiteral() {
        return this.getToken(CashScriptParser.BooleanLiteral, 0);
    }
    numberLiteral() {
        return this.getTypedRuleContext(NumberLiteralContext, 0);
    }
    StringLiteral() {
        return this.getToken(CashScriptParser.StringLiteral, 0);
    }
    DateLiteral() {
        return this.getToken(CashScriptParser.DateLiteral, 0);
    }
    HexLiteral() {
        return this.getToken(CashScriptParser.HexLiteral, 0);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_literal;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitLiteral) {
            return visitor.visitLiteral(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class NumberLiteralContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    NumberLiteral() {
        return this.getToken(CashScriptParser.NumberLiteral, 0);
    }
    NumberUnit() {
        return this.getToken(CashScriptParser.NumberUnit, 0);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_numberLiteral;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitNumberLiteral) {
            return visitor.visitNumberLiteral(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
export class TypeNameContext extends ParserRuleContext {
    constructor(parser, parent, invokingState) {
        super(parent, invokingState);
        this.parser = parser;
    }
    Bytes() {
        return this.getToken(CashScriptParser.Bytes, 0);
    }
    get ruleIndex() {
        return CashScriptParser.RULE_typeName;
    }
    // @Override
    accept(visitor) {
        if (visitor.visitTypeName) {
            return visitor.visitTypeName(this);
        }
        else {
            return visitor.visitChildren(this);
        }
    }
}
