// cashc/dist/artifact/Artifact.js
import { scriptToAsm, } from '@cashscript/utils';
import { version } from '../index.js';
export function generateArtifact(ast, script, source, debug) {
    const { contract } = ast;
    const constructorInputs = contract.parameters
        .map((parameter) => ({ name: parameter.name, type: parameter.type.toString() }));
    const abi = contract.functions.map((func) => ({
        name: func.name,
        inputs: func.parameters.map((parameter) => ({
            name: parameter.name,
            type: parameter.type.toString(),
        })),
    }));
    const bytecode = scriptToAsm(script);
    return {
        contractName: contract.name,
        constructorInputs,
        abi,
        bytecode,
        source,
        debug,
        compiler: {
            name: 'cashc',
            version,
        },
        updatedAt: new Date().toISOString(),
    };
}
// strip verbose debug info from artifact for production use
export function stripArtifact(artifact) {
    delete artifact.debug;
    return artifact;
}
