// cashc/dist/compiler.js
import { CharStream, CommonTokenStream } from 'antlr4';
import { binToHex } from '@bitauth/libauth';
import { generateSourceMap, optimiseBytecode, optimiseBytecodeOld, scriptToAsm, scriptToBytecode, sourceMapToLocationData } from '@cashscript/utils';
import fs from 'fs';
import { generateArtifact } from './artifact/Artifact.js';
import AstBuilder from './ast/AstBuilder.js';
import ThrowingErrorListener from './ast/ThrowingErrorListener.js';
import GenerateTargetTraversal from './generation/GenerateTargetTraversal.js';
import CashScriptLexer from './grammar/CashScriptLexer.js';
import CashScriptParser from './grammar/CashScriptParser.js';
import SymbolTableTraversal from './semantic/SymbolTableTraversal.js';
import TypeCheckTraversal from './semantic/TypeCheckTraversal.js';
import EnsureFinalRequireTraversal from './semantic/EnsureFinalRequireTraversal.js';
export function compileString(code) {
    // Lexing + parsing
    let ast = parseCode(code);
    // Semantic analysis
    ast = ast.accept(new SymbolTableTraversal());
    ast = ast.accept(new TypeCheckTraversal());
    ast = ast.accept(new EnsureFinalRequireTraversal());
    // Code generation
    const traversal = new GenerateTargetTraversal();
    ast = ast.accept(traversal);
    const constructorParamLength = ast.contract.parameters.length;
    // Bytecode optimisation
    const optimisedBytecodeOld = optimiseBytecodeOld(traversal.output);
    const optimisationResult = optimiseBytecode(traversal.output, sourceMapToLocationData(traversal.sourceMap), traversal.consoleLogs, traversal.requires, constructorParamLength);
    if (scriptToAsm(optimisedBytecodeOld) !== scriptToAsm(optimisationResult.script)) {
        console.error(scriptToAsm(optimisedBytecodeOld));
        console.error(scriptToAsm(optimisationResult.script));
        throw new Error('New bytecode optimisation is not backwards compatible, please report this issue to the CashScript team');
    }
    // Attach debug information
    const debug = {
        bytecode: binToHex(scriptToBytecode(optimisationResult.script)),
        sourceMap: generateSourceMap(optimisationResult.locationData),
        logs: optimisationResult.logs,
        requires: optimisationResult.requires,
    };
    return generateArtifact(ast, optimisationResult.script, code, debug);
}
export function compileFile(codeFile) {
    const code = fs.readFileSync(codeFile, { encoding: 'utf-8' });
    return compileString(code);
}
export function parseCode(code) {
    // Lexing (throwing on errors)
    const inputStream = new CharStream(code);
    const lexer = new CashScriptLexer(inputStream);
    lexer.removeErrorListeners();
    lexer.addErrorListener(ThrowingErrorListener.INSTANCE);
    const tokenStream = new CommonTokenStream(lexer);
    // Parsing (throwing on errors)
    const parser = new CashScriptParser(tokenStream);
    parser.removeErrorListeners();
    parser.addErrorListener(ThrowingErrorListener.INSTANCE);
    const parseTree = parser.sourceFile();
    // AST building
    const ast = new AstBuilder(parseTree).build();
    return ast;
}
